godot-ffi/src/binding/single_threaded.rs

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright (c) godot-rust; Bromeon and contributors.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+
+//! Non-thread safe binding storage.
+//!
+//! If used from different threads then there will be runtime errors in debug mode and UB in release mode.
+
+use std::{cell::OnceCell, sync::OnceLock, thread::ThreadId};
+
+use crate::{
+    BuiltinLifecycleTable, BuiltinMethodTable, ClassEditorMethodTable, ClassSceneMethodTable,
+    ClassServersMethodTable, GDExtensionInterface, GdextRuntimeMetadata, UtilityFunctionTable,
+};
+
+use super::GDExtensionClassLibraryPtr;
+
+pub(super) struct BindingStorage {
+    // Is used in to check that we've been called from the right thread, so must be thread-safe to access.
+    main_thread: OnceLock<ThreadId>,
+    binding: OnceCell<GodotBinding>,
+}
+
+impl BindingStorage {
+    /// Get the static binding storage.
+    ///
+    /// # Safety
+    ///
+    /// You must not access `binding` from a thread different than the thread [`initialize`](BindingStorage::initialize) was first called from.
+    #[inline(always)]
+    unsafe fn storage() -> &'static Self {
+        static BINDING: BindingStorage = BindingStorage {
+            main_thread: OnceLock::new(),
+            binding: OnceCell::new(),
+        };
+
+        &BINDING
+    }
+
+    /// Initialize the binding storage, this must be called before any other public functions.
+    #[must_use]
+    pub fn initialize(binding: GodotBinding) -> Option<()> {
+        // SAFETY: Either we are the first call to `initialize` and so we are calling from the same thread as ourself. Or we are a later call,
+        // in which case we can tell that the storage has been initialized and dont access `binding`.
+        let storage = unsafe { Self::storage() };
+
+        storage.main_thread.set(std::thread::current().id()).ok()?;
+        storage
+            .binding
+            .set(binding)
+            .ok()
+            .expect("`main_thread` was unset so `binding` should also be unset");
+
+        Some(())
+    }
```

Comment:
Might be more idiomatic to return `Result<(), ()>` here? Would also not need extra `#[must_use]` and `.ok()` calls, and directly expresses that the return type is success/failure, and not just something optional.

Otherwise we could as well use `bool`...